llama_index/embeddings/dashscope.py

```diff
@@ -0,0 +1,331 @@
+        https://help.aliyun.com/zh/dashscope/developer-reference/text-embedding-async-api-details
+    """
+    try:
+        import dashscope
+    except ImportError:
+        raise ImportError("DashScope requires `pip install dashscope")
+    response = dashscope.MultiModalEmbedding.call(
+        model=model, input=input, api_key=api_key, kwargs=kwargs
+    )
+    if response.status_code == HTTPStatus.OK:
+        return response.output["embedding"]
+    else:
+        logger.error("Calling MultiModalEmbedding failed, details: %s" % response)
+        return []
+
+
+class DashScopeTextEmbedding(BaseEmbedding):
+    """DashScope class for text embedding.
+
+    Args:
+        model (str): Model for embedding.
+            Defaults to DashScopeTextEmbeddingModels.TEXT_EMBEDDING_V2.
+                Options are:
+
+                - DashScopeTextEmbeddingModels.TEXT_EMBEDDING_V1
+                - DashScopeTextEmbeddingModels.TEXT_EMBEDDING_V2
+        api_key (str): The DashScope api key.
+    """
+
+    _api_key: Optional[str] = PrivateAttr()
+
+    def __init__(
+        self,
+        model: str = DashScopeTextEmbeddingModels.TEXT_EMBEDDING_V2,
+        api_key: Optional[str] = None,
+        **kwargs: Any,
+    ) -> None:
+        self._api_key = api_key
+        super().__init__(
+            model_name=model,
+            **kwargs,
+        )
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "DashScopeEmbedding"
+
+    def _get_query_embedding(self, query: str) -> List[float]:
+        """Get query embedding."""
+        emb = get_text_embedding(
+            self.model_name,
+            query,
+            api_key=self._api_key,
+            text_type=DashScopeTextEmbeddingType.TEXT_TYPE_QUERY,
+        )
+        if len(emb) > 0:
+            return emb[0]
+        else:
+            return []
+
+    def _get_text_embedding(self, text: str) -> List[float]:
+        """Get text embedding."""
+        emb = get_text_embedding(
+            self.model_name,
+            text,
+            api_key=self._api_key,
+            text_type=DashScopeTextEmbeddingType.TEXT_TYPE_DOCUMENT,
+        )
+        if len(emb) > 0:
+            return emb[0]
+        else:
+            return []
+
+    def _get_text_embeddings(self, texts: List[str]) -> List[List[float]]:
+        """Get text embeddings."""
+        return get_text_embedding(
+            self.model_name,
+            texts,
+            api_key=self._api_key,
+            text_type=DashScopeTextEmbeddingType.TEXT_TYPE_DOCUMENT,
+        )
+
+    # TODO: use proper async methods
+    async def _aget_text_embedding(self, query: str) -> List[float]:
+        """Get text embedding."""
+        return self._get_text_embedding(query)
+
+    # TODO: user proper async methods
+    async def _aget_query_embedding(self, query: str) -> List[float]:
+        """Get query embedding."""
+        return self._get_query_embedding(query)
+
+
+class DashScopeBatchTextEmbedding:
```

Comment:
It seems like this class is unused/uneeded?

Reply:
This is also an embedding service, but the interface is different from the online interface. It focuses on offline processing of large amounts of data. Input is the file to be embedding (one per line) url, and after the processing is completed, the result file URL is returned.